hw04-logger/src/main/java/ru/otus/hw04/DI.java

```diff
@@ -0,0 +1,84 @@
+package ru.otus.hw04;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.Proxy;
+import java.util.*;
+
+public class DI {
+    public OriginalInterface createProxyFromOriginal(Object original) {
+        Class clazz = original.getClass();
+        ClassLoader classLoader = clazz.getClassLoader();
+        Class<?>[] interfaces = clazz.getInterfaces();
+        CustomInvocationHandler invocationHandler = new CustomInvocationHandler(original);
+
+        return (OriginalInterface) Proxy.newProxyInstance(classLoader, interfaces, invocationHandler);
+    }
+
+    class CustomInvocationHandler implements InvocationHandler {
+        private Object original;
+        private Set<String> methodsNeedLogging = new HashSet<>();
+
+        /**
+         * @param original
+         */
+        public CustomInvocationHandler(Object original) {
+
+            this.original = original;
+
+            Class thisClass = original.getClass();
+            Method[] methods = thisClass.getDeclaredMethods();
+            for (Method method : methods) {
+                if (method.getAnnotation(Log.class) instanceof Log) {
```

Comment:
```suggestion
                if (method.getAnnotation(Log.class) != null) {
```

Если `@Log` нет, то будет null.

Reply:
или так `method.isAnnotationPresent(Log.class)`